### PR TITLE
RFC: Improve consistency of Date formatting

### PR DESCRIPTION
**Pull request by omus**
After the work on #15224 had completed I came up with an alternative that eliminates the need for the 'Y' slot and makes the slots work in a more consistent manner. Take the example of formatting a month as a string:

``` julia
julia> Dates.format(DateTime(1996, 4), "m")
"4"

julia> Dates.format(DateTime(1996, 4), "mm")
"04"

julia> Dates.format(DateTime(1996, 12), "m")
"12"

julia> Dates.format(DateTime(1996, 12), "mm")
"12"
```

When the number of digits in the month exceeds the width of the slot then all of the digits of the month are returned. Contrast this with year:

``` julia
julia> Dates.format(DateTime(1996, 4), "y")
"6"

julia> Dates.format(DateTime(1996, 4), "yy")
"96"
```

This PR attempts to unify these behaviours by only truncating a slot when the width is two or greater:

``` julia
julia> Dates.format(DateTime(1996, 4), "y m")  # New breaking behaviour
"1996 4"

julia> Dates.format(DateTime(1996, 4), "yy mm")
"96 04"
```

Another issue which this PR addresses is how milliseconds where always assumed to be preceded by a period:

``` julia
julia> Dates.format(DateTime(1,1,1,0,0,0,250), "s")
"25"

julia> Dates.format(DateTime(1,1,1,0,0,0,250), "ss")
"25"

julia> Dates.format(DateTime(1,1,1,0,0,0,250), ".s")
".25"
```

The code now looks at the preceeding delimiter and adjusts the behaviour accordingly:

``` julia
julia> Dates.format(DateTime(1,1,1,0,0,0,250), "s")  # New behaviour
"250"

julia> Dates.format(DateTime(1,1,1,0,0,0,250), "ss")  # New behaviour
"50"

julia> Dates.format(DateTime(1,1,1,0,0,0,250), ".s")
".25"
```

To summarize:
- Changes now make working with formatting more consistent
- `y` slot with a width of one returns the entire year rather than just the last digit
- `Y` slot removed
- `s` (millisecond) slot behaviour changes when preceded by a period

### CODE REVIEW

**Comment by omus** [comment]
cc: @quinnj

**Comment by StefanKarpinski** [comment]
I can see that this accomplishes what people generally want done, but it feels so very ad hoc. I kind of prefer just introducing the `Y` slot that doesn't truncate. What is the current millisecond behavior?

**Comment by iamed2** [comment]
`Y` feels like the more ad hoc solution to me. The new `y` approach follows with what `m` for months and `d` for days do.

The old millisecond behaviour (it's unaware of surrounding characters):
```
julia> Dates.format(DateTime(1,1,1,0,0,0,250), "s")
"25"

julia> Dates.format(DateTime(1,1,1,0,0,0,250), "ss")
"25"

julia> Dates.format(DateTime(1,1,1,0,0,0,250), "sss")
"250"

julia> Dates.format(DateTime(1,1,1,0,0,0,250), "ssss")
"2500"
```

I think a whole new DateFormat approach is needed (maybe printf-style or python string format-style with options applied to the fields) but that this is the best way of addressing the current issues without rewriting the whole thing.

**Comment by omus** [comment]
If we want to start needing more advanced features then we could move to a printf style format. At this point though I don't think that kind of change is needed.

I'd also like to point out that the old millisecond behaviour and the old year behaviour are almost opposites of each other. Millisecond right pads zeros while year left pads:
```julia
julia> Dates.format(DateTime(1996), "y")
"6"

julia> Dates.format(DateTime(1996), "yy")
"96"

julia> Dates.format(DateTime(1996), "yyy")
"996"

julia> Dates.format(DateTime(1996), "yyyy")
"1996"

julia> Dates.format(DateTime(1996), "yyyyy")
"01996"
```

**Comment by omus** [comment]
Rebased the changes. I also discovered a pre-existing bug. I'm hoping that this can be merged prior to release 0.5 as it would mean the `'Y'` slot would have only have existed in Julia 0.5-dev

File: test/dates/io.jl [hunk 1]
```diff
@@ -359,3 +417,38 @@ let
...
+    @test Dates.format(dt, "yyyyyy") == "001991"
+    @test Dates.format(dt, "m") == "2"
+    @test Dates.format(dt, "mm") == "02"
+    @test Dates.format(dt, "mmm") == "002"
+
+    @test Dates.format(dt, "s") == "7"
+    @test Dates.format(dt, "ssss") == "0007"
+    @test Dates.format(dt, ".s") == ".007"
+    @test Dates.format(dt, ".ssss") == ".0070"
+
```
Comment (tkelman) [inline test/dates/io.jl]: unneeded

otherwise looks okay to me, any performance implications? is there a set of dates benchmarks in the suite that nanosoldier runs yet?

Comment (iamed2) [inline test/dates/io.jl]: Why do you believe these are not useful?

Comment (tkelman) [inline test/dates/io.jl]: just the blank line, minor style nit

Comment (omus) [inline test/dates/io.jl]: Taken care of

Comment (tkelman) [inline test/dates/io.jl]: still there?

**Comment by vtjnash** [comment]
It seems like the bug in the last commit was fixed. Was there anything else still needed from here? I am cleanup up some possibly stale PRs.
```
julia> @test Dates.Date("2009(年)12(月)01(日)","yyyy(年)mm(月)dd(日)") == Dates.Date(2009,12,1)
Test Passed

julia> @test Dates.format(Dates.Date(2009,12,1),"yyyy(年)mm(月)dd(日)") == "2009(年)12(月)01(日)"
Test Passed
```

**Comment by omus** [comment]
I had completely forgotten about this PR. I'll review if anything else is needed from here.
